File: cs/cs-kata/Bowling180630.cs
using System;
using System.Collections.Generic;

namespace cs_kata
{
    public class Bowling180630
    {
        private List<int> rolls = new List<int>();

        public void Roll(int pins)
        {
            this.rolls.Add(pins);
        }
        
        public int Score() 
        {
            int score = 0;
            int rollIndex = 0;
            for (int frame = 0; frame < 10; frame ++) {
                if (this.IsStrike(rollIndex)) {
                    score += this.StrikeScore(rollIndex);
                    rollIndex += 1;
                } else if (this.IsSpare(rollIndex)) {
                    score += this.SpareScore(rollIndex);
                    rollIndex += 2;
                } else {
                    score += this.NormalFrameScore(rollIndex);
                    rollIndex += 2;
                }
            }
            return score;
        }

        private bool IsStrike(int rollIndex)
        {
            return 10 == this.rolls[rollIndex];
        }

        private bool IsSpare(int rollIndex)
        {
            return 10 == (this.rolls[rollIndex] + this.rolls[rollIndex + 1]);
        }

        private int StrikeScore(int rollIndex)
        {
            return 10 + this.rolls[rollIndex + 1] + this.rolls[rollIndex + 2];
        }

        private int SpareScore(int rollIndex)
        {
            return 10 + this.rolls[rollIndex + 2];
        }

        private int NormalFrameScore(int rollIndex)
        {
            return this.rolls[rollIndex] + this.rolls[rollIndex + 1];
        }
    }
}

File: go/180630_bowling/bowling_180630.go
package main

func main() {

}

// Bowling180630 ...
type Bowling180630 struct {
	rolls []int
}

func (b *Bowling180630) roll(pins int) {
	b.rolls = append(b.rolls, pins)
}

func (b Bowling180630) score() int {
	var score = 0
	var rollIndex = 0
	for frame := 0; frame < 10; frame++ {
		if b.isStrike(rollIndex) {
			score += b.strikeScore(rollIndex)
			rollIndex++
		} else if b.isSpare(rollIndex) {
			score += b.spareScore(rollIndex)
			rollIndex += 2
		} else {
			score += b.normalFrameScore(rollIndex)
			rollIndex += 2
		}
	}
	return score
}

func (b Bowling180630) isStrike(rollIndex int) bool {
	return 10 == b.rolls[rollIndex]
}

func (b Bowling180630) isSpare(rollIndex int) bool {
	return 10 == (b.rolls[rollIndex] + b.rolls[rollIndex+1])
}

func (b Bowling180630) strikeScore(rollIndex int) int {
	return 10 + b.rolls[rollIndex+1] + b.rolls[rollIndex+2]
}

func (b Bowling180630) spareScore(rollIndex int) int {
	return 10 + b.rolls[rollIndex+2]
}

func (b Bowling180630) normalFrameScore(rollIndex int) int {
	return b.rolls[rollIndex] + b.rolls[rollIndex+1]
}
